botocore/translate.py

```diff
@@ -11,22 +11,46 @@
 # distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
 # ANY KIND, either express or implied. See the License for the specific
 # language governing permissions and limitations under the License.
+import copy
+
 from botocore.utils import merge_dicts
 
 
-def build_retry_config(endpoint_prefix, retry_model, definitions):
+def build_retry_config(endpoint_prefix, retry_model, definitions,
+                       client_retry_config=None):
     service_config = retry_model.get(endpoint_prefix, {})
     resolve_references(service_config, definitions)
     # We want to merge the global defaults with the service specific
     # defaults, with the service specific defaults taking precedence.
     # So we use the global defaults as the base.
-    final_retry_config = {'__default__': retry_model.get('__default__', {})}
+    #
+    # A deepcopy is done on the retry defaults because it ensures the
+    # retry model has no chance of getting mutated when the service specific
+    # configuration or client retry config is merged in.
+    final_retry_config = {
+        '__default__': copy.deepcopy(retry_model.get('__default__', {}))
+    }
     resolve_references(final_retry_config, definitions)
     # The merge the service specific config on top.
     merge_dicts(final_retry_config, service_config)
+    if client_retry_config is not None:
+        _merge_client_retry_config(final_retry_config, client_retry_config)
     return final_retry_config
 
 
+def _merge_client_retry_config(retry_config, client_retry_config):
+    max_retry_attempts_override = client_retry_config.get('max_attempts')
+    if max_retry_attempts_override is not None:
+        # In the retry config, the max_attempts refers to the maximum number
+        # of requests in general will be made. However, for the client's
+        # retry config it refers to how many retry attempts will be made at
+        # most. So to translate this number from the client config, one is
+        # added to convert it to the maximum number request that will be made
+        # by including the initial request.
+        retry_config['__default__'][
```

Comment:
We don't support this now, but if we ever support `max_attempts` at the operation level, we'll need to revisit how we do this.  Might be worth adding a note about that.